### PR TITLE
perf(acc-hashing): use rayon threadpool to generate hashed accounts

### PR DESCRIPTION
**Pull request by gakonst**
Re-use the same technique from the SenderRecovery stage: https://github.com/paradigmxyz/reth/blob/724ec1b5cc4148029181d2d29c4f29e57e6b3878/crates/stages/src/stages/sender_recovery.rs#L90-L114

`cargo bench` is showing a regression in performance, interestingly enough. Maybe because the bench is small and Rayon overhead dominates? I'm gonna run this in prod and see what comes back.

Also, if we want to use the sequential stream, we'd need to make modifications to it so that it can tiebreak on keys that are not Add/One, see below

```patch
diff --git a/crates/stages/src/stages/hashing_account.rs b/crates/stages/src/stages/hashing_account.rs
index 9956bee2..2f000042 100644
--- a/crates/stages/src/stages/hashing_account.rs
+++ b/crates/stages/src/stages/hashing_account.rs
@@ -226,8 +226,13 @@ impl<DB: Database> Stage<DB> for AccountHashingStage {
                 // drop the sender to signal that we're done sending
                 drop(sender);
 
-                let mut hashed_accounts = UnboundedReceiverStream::new(rx);
+                // We sort everything inside the batch to minimize the number of cursor seeks
+                let mut hashed_accounts = SequentialPairStream::new(
+                    keccak256(start_address.unwrap_or_default()),
+                    UnboundedReceiverStream::new(rx),
+                );
                 let mut hashed_account_cursor = tx.cursor_write::<tables::HashedAccount>()?;
+
                 while let Some(account) = hashed_accounts.next().await {
                     let (k, v) = account?;
                     hashed_account_cursor.insert(k, v)?;
diff --git a/crates/stages/src/stages/stream.rs b/crates/stages/src/stages/stream.rs
index 5e8f41e3..408091bc 100644
--- a/crates/stages/src/stages/stream.rs
+++ b/crates/stages/src/stages/stream.rs
@@ -23,7 +23,11 @@ pub(crate) struct SequentialPairStream<Key: Ord, Value, St> {
 
 // === impl SequentialPairStream ===
 
-impl<Key: Ord, Value, St> SequentialPairStream<Key, Value, St> {
+impl<Err, Key: Ord, Value, St> SequentialPairStream<Key, Value, St>
+where
+    Key: Ord + Copy,
+    St: Stream<Item = Result<(Key, Value), Err>>,
+{
     /// Returns a new [SequentialPairStream] that emits the items of the given stream in order
     /// starting at the given start point.
     pub(crate) fn new(start: Key, stream: St) -> Self {
@@ -34,7 +38,7 @@ impl<Key: Ord, Value, St> SequentialPairStream<Key, Value, St> {
 /// implements Stream for any underlying Stream that returns a result
 impl<Key, Value, St, Err> Stream for SequentialPairStream<Key, Value, St>
 where
-    Key: Ord + Copy + Add<Output = Key> + One,
+    Key: Ord + Copy,
     St: Stream<Item = Result<(Key, Value), Err>>,
 {
     type Item = Result<(Key, Value), Err>;
@@ -52,7 +56,7 @@ where
                     }
                     Ordering::Equal => {
                         let next = PeekMut::pop(maybe_next);
-                        *this.next = *this.next + Key::one();
+                        // *this.next = *this.next + Key::one();
                         return Poll::Ready(Some(Ok(next.into())))
                     }
                     Ordering::Greater => {
@@ -75,7 +79,7 @@ where
                     Poll::Ready(item) => match item {
                         Some(Ok((k, v))) => {
                             if k == *this.next {
-                                *this.next = *this.next + Key::one();
+                                // *this.next = *this.next + Key::one();
                                 return Poll::Ready(Some(Ok((k, v))))
                             }
                             this.pending.push(OrderedItem(k, v));
```

Here's the flamegraph on master :
flame-main

And here's the flamegraph on this PR:
flamegraph

Both produced via:
```
cargo b --profile debug-fast
flamegraph -- ./target/debug-fast/reth node --debug.max-block <...> --metrics loccalhost
```

Honestly I'm not exactly sure. The gap between the two flamegraphs is the checkpoint, and some p2p logic running alongside it. There's also a big _clone thing that takes 10% of the time. The flamegraphs are going to be easier to parse if you download the file and open them locally in your svg viewer/browser (vs from the github hosted one)

### CODE REVIEW

**Comment by gakonst** [comment]
if this approach ends up working well, we should do the same for storage hashing

File: crates/stages/src/stages/hashing_account.rs [hunk 1]
```diff
@@ -190,27 +194,39 @@ impl<DB: Database> Stage<DB> for AccountHashingStage {
...
+            let (sender, rx) = mpsc::unbounded_channel();
             let start_address = checkpoint.address.take();
             let next_address = {
                 let mut accounts = tx.cursor_read::<tables::PlainAccountState>()?;
 
-                let hashed_batch = accounts
-                    .walk(start_address)?
-                    .take(self.commit_threshold as usize)
-                    .map(|res| res.map(|(address, account)| (keccak256(address), account)))
-                    .collect::<Result<BTreeMap<_, _>, _>>()?;
```
Comment (rakita) [inline crates/stages/src/stages/hashing_account.rs:201]: It is probably bcs of this. We collect hashed to BTreeMap and insert to the database in a sorted way.

There is probably some rayon helper that would iterate and transform values for us, and after the transformation is done, we sort it and insert it inside the db.

**Comment by gakonst** [comment]
@joshieDo @rakita @mattsse Did we basically decide that any of these optimizations are not really worth doing because we're I/O bound when reading in plainstate?

**Comment by mattsse** [comment]
> Did we basically decide that any of these optimizations are not really worth doing 

can't really comment on this because not familiar with this stage.
The rayon usage in the sender recovery stage has since been changed and is disk IO bound.

**Comment by rakita** [comment]
superseded by this PR: https://github.com/paradigmxyz/reth/pull/2077

**Pull request closed without merging**